Add option to compress BSA

dmdemoura opened the pull request:
I runned into an issue where, I think, Skyrim SE didn't work with uncompressed textures BSA.
So I added an extra option to the Package key, that lets you toggle BSA compression.

Example:
```xml
<Packages Output="@Data">
    <Package Name="@ModName" RootDir="@CKData">
        <Include>@CKData\Scripts\*.pex</Include>
        <Include>@CKData\Meshes\**\*.nif</Include>
        <Include>@CKData\Meshes\**\*.tri</Include>
        <Include>Seq\@ModName.seq</Include>
        <Include>@CKData\Sound\Voice\@ModName.esp\**\*.fuz</Include>
    </Package>
    <Package Name="@ModName - Textures" RootDir="@CKData" Compress="True">
        <Include>@CKData\Textures\**\*.dds</Include>
    </Package>
</Packages>
```

fireundubh commented:
Thanks for the PR.

According to Zilav (author of BSArch), uncompressed textures are fine, except when the BSA has the Embed Filenames flag. This is only a bug in SSE. BSArch automatically adds the Embed Filenames flag to DDS-only archives.

The simpler solution would be to pass `-af:0x3` to BSArch if any DDS files are queued to be packed. Can you write and test that?

dmdemoura commented:
Hum, interesting. I'll try implementing that once I have some more time :D

dmdemoura commented:
So I implemented the check to add -af:0x3 to SSE BSAs that contain any DDS files.
I retested the BSA options with SkyrimSE, and can confirm that:
- Uncompressed textures bsa with EmbedNames flag = CTD.
- Uncompressed textures bsa without EmbedNames flag = Works fine.
- Compressed textures bsa with EmbedNames flag = Also works fine.

I implemented the EmbedName fix on a separate branch, so it's possible for me to pull request just that if needed.

dmdemoura commented on pyro/PackageManager.py:
```diff
@@ -184,7 +184,7 @@ def create_packages(self) -> None:
                 shutil.copy2(source_path, target_path)
 
-            compress: bool = package_node.get('Compress')
+            compress: bool = package_node.get('Compress') == 'True'
```
Here I found a bug I made, that caused Compress="False" to enable compression.
It's fixed now, I copied this line from here: https://github.com/fireundubh/pyro/blob/f4fd061609e8db44c7c51014240a6220a64e500d/pyro/PapyrusProject.py#L96

But this means that only Compress="True" evaluates to True, even though the XML schema suggests that true, tRue, and 1 should also work: https://github.com/fireundubh/pyro/blob/f4fd061609e8db44c7c51014240a6220a64e500d/pyro/PapyrusProject.xsd#L115-L120

Is this a bug or am I confused?

fireundubh replied:
> Is this a bug or am I confused?

The first two patterns in the XSD allow users to enter `true` or `false` case insensitively.

The third pattern allows `1` and `0` whose meanings are handled in Python.

`PapyrusProject._update_attributes` cleans up, formats, and/or defaults relevant attribute values, and `str(True) == 'True'` in Python. If you add a bool key, you need to declare it in one of the lists at the top of this method. The loop at the bottom will ensure the bool can be referenced properly.

So, by the time we call `self.ppj_root.get('Optimize')`, the value is either `'True'` or `'False'` and cannot be anything else.

dmdemoura replied:
Okay, now it makes sense :D. Wasn't working for compress because I hadn't added it to _update_attributes.

fireundubh commented:
I'd rather not add a `Compress` attribute. The `store` compression type ("compression" type) already means `uncompressed`.

I'd also rather not change the signature of `PackageManager.build_commands` for this workaround.

How about we go with this?

```python
        if self.options.game_type == GameType.FO4:
            arguments.append('-fo4')
        elif self.options.game_type == GameType.SSE:
            arguments.append('-sse')
            
            # SSE has an ctd bug with uncompressed textures in a bsa that
            # has an Embed Filenames flag on it, so force it to false.
            has_textures = False

            for f in glob.iglob(os.path.join(containing_folder, r'**/*'), recursive=True):
                if not os.path.isfile(f):
                    continue
                if endswith(f, '.dds', ignorecase=True):
                    has_textures = True
                    break

            if has_textures:
                arguments.append('-af:0x3')
        else:
            arguments.append('-tes5')
```

If Bethesda ever fixes their bug, this override will be easy to remove.

dmdemoura commented:
It looks good, and as you say it does minimize the amount of changes for what is essentially a bug workaround.

dmdemoura commented:
I tested your solution and it does work perfectly. The pull request was updated to match it.
